episodes/3-documentation.md

```diff
@@ -216,49 +215,148 @@ def fahrenheit_to_celsius(temp_f: float) -> float:
+
 Docstrings are more powerful than comments:
 
-- Docstrings are automatically extracted when calling the help function.
-- Tools can generate documentation pages automatically from docstrings known as API documentation.
+- Docstrings are automatically extracted when calling the `help` function.
+- Tools (e.g. `mkdocs`, `sphinx`) can generate documentation pages automatically from docstrings known as API documentation.
 
-It is common to write docstrings for functions, classes, and modules.
+A good docstring should describe:
 
-TODO: Introduce how to write docstrings and show an example of API documentation.
-Good docstrings describe:
 - What the function does
 - What goes in (including the type of the input variables)
 - What goes out (including the return type)
 
+There are different docstrings styles, such as Google style, numpy style and reStructered style, with Google style being the most common.
+
+::::::::::::::::::::::::::::::::::::: challenge 
+
+## Exercise: Write a docstring 
+
+``` python
+def check_temperature_validity(temperature, unit):
+    abs_zero = {"C": -273.15, "F": -459.67, "K": 0}
+    if temperature < abs_zero[unit]:
+        return False
+    return True
+
+```
+:::::::::::::::::::::::: solution 
+
+## Solution
+
+Caveat: without specifying the expected data types of function arguments and return values it is unclear what these values should be! This is called _type hinting_. Lets add types and write the docstring.
+
+``` python
+def check_temperature_validity(temperature: float, unit: str) -> bool:
+    """
+    Checks if a temperature is valid for a given unit.
+
+    Args:
+        temperature (float): The temperature to check.
+        unit (str): The unit of the temperature. Must be "C", "F", or "K".
+
+    Returns:
+        bool: True if the temperature is valid, False otherwise.
+    """
+    abs_zero = {"C": -273.15, "F": -459.67, "K": 0}
+    if temperature < abs_zero[unit]:
+        return False
+    return True
+```
+:::::::::::::::::::::::: 
+
+:::::::::::::::::::::::::::::::::::::
+
+
+::::::::::::::::::::::::::::::::::::: challenge 
+## Exercise: Identify proper docstring
+
+Which of the following options is the correct function with a proper docstring?
+
+**Option A**
+
+```python
+def check_unit_validity(unit: str) -> bool:
+    """
+    Checks if a unit is valid.
+
+    Args:
+        unit (str): The unit to check. Must be "C", "F", or "K".
+
+    Returns:
+        bool: True if the unit is valid, False otherwise.
+    """
+    if not unit in ["C", "F", "K"]:
+        return False
+    return True
+```
+
+**Option B**
+
+```python
+def check_unit_validity(unit: int) -> bool:
+    """
+    Checks if a unit is valid.
+
+    Args:
+        unit (int): The unit to check. Must be "C", "F", or "K".
+
+    Returns:
+        bool: True if the unit is valid, False otherwise.
+    """
+    if not unit in ["C", "F", "K"]:
+        return False
+    return True
+```
```

Comment:
I think this is a bit lame, because it is more a mistake than an incomplete docstring. I would suggest to just leave out the types:

```suggestion
def check_unit_validity(unit):
    """
    Checks if a unit is valid.

    Args:
        unit: The unit to check. Must be "C", "F", or "K".

    Returns:
        bool: True if the unit is valid, False otherwise.
    """
    if not unit in ["C", "F", "K"]:
        return False
    return True
```
```

Reply:
I tweaked the exercise! See if you like it now.